### ISSUE TITLE
Serve `.gz` files on GitHub Pages with the proper `Content-Encoding: gzip` header

### ISSUE THREAD

**Issue by mathiasbynens** [isaacs/github#611]
> Dear GitHub Support,
> 
> Please serve precompressed `.gz` files on GitHub Pages with the appropriate `Content-Encoding: gzip` header.
> 
> The `Content-Type` header is already correctly set to `application/octet-stream` (see [https://mail-archives.apache.org/mod_mbox/httpd-dev/200207.mbox/raw/<3D2D4E76.4010502@talex.com.pl>/](https://mail-archives.apache.org/mod_mbox/httpd-dev/200207.mbox/raw/<3D2D4E76.4010502@talex.com.pl>/)). All that’s left is the `Content-Encoding` header.
> 
> Current output:
> 
> ```
> $ curl -I https://tibiamaps.github.io/tibia-map-data/mapper/12712707.map.gz
> HTTP/1.1 200 OK
> Server: GitHub.com
> Content-Type: application/octet-stream
> Last-Modified: Tue, 15 Mar 2016 23:46:34 GMT
> Access-Control-Allow-Origin: *
> Expires: Wed, 16 Mar 2016 09:08:27 GMT
> Cache-Control: max-age=600
> X-GitHub-Request-Id: 17EB2B1E:0345:370F9E6:56E92033
> Content-Length: 9014
> Accept-Ranges: bytes
> Date: Wed, 16 Mar 2016 09:16:53 GMT
> Via: 1.1 varnish
> Age: 95
> Connection: keep-alive
> X-Served-By: cache-ams4142-AMS
> X-Cache: HIT
> X-Cache-Hits: 1
> X-Timer: S1458119813.598808,VS0,VE0
> Vary: Accept-Encoding
> X-Fastly-Request-ID: 31363c9a270ac13e51cf73d8b822435d0b6d5d96
> ```
> 
> Note that `Content-Encoding: gzip` is missing. Please add it.
> 
> This is currently breaking an application I’m working on. It seems weird to punish users for compressing their GitHub Pages-hosted content and saving your bandwidth.
> 
> Thanks for considering,
> Mathias

**Comment by TazeTSchnitzel** [comment]
This presupposes that users always want `.gz` to be used only to transparently compress content in transit, rather than wanting to download a gzip file.

The former is what `Content-Encoding` does, e.g. with [nginx's `gzip_static`](https://t.co/a6binSS1mN), where you can add a pre-compressed `.gz` version of a file (say `win95.data.gz`), and then when a user requests the file without the `.gz` extension (e.g. `win95.data`), the content of the `.gz` file is sent to the browser with a `Content-Encoding: gzip` header. The browser decompresses the file while it downloads, and the user gets an uncompressed file. To test this out in the real world, download http://win95.ajf.me/win95.data and look at the Network tab. Your browser will download a 49MB gzipped file over the network, but the file it'll actually save to disk is 131MB, because the `Content-Encoding: gzip` header told it to decompress it.

**Comment by mhart** [comment]
`Content-Encoding` is supposed to be used when you have a `Content-Type` that you want the browser to display/intepret, but you've encoded/compressed it. Eg, a gzip HTML file with a `Content-Type` of `text/html`.

From [the spec](https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html):

> When present, its value indicates what additional content codings have been applied to the entity-body, and thus what decoding mechanisms must be applied in order to obtain the media-type referenced by the Content-Type header field.

**Comment by TazeTSchnitzel** [comment]
If GitHub added this header to .gz files automatically, it'd mean that .tar.gz files became .tar files when downloaded.

**Comment by charliesome** [comment]
In this case, omitting `Content-Encoding` is required for correct behaviour.

Browsers interpret the `Content-Encoding` header as describing what additional encoding has been applied to the actual content of the response when writing out the body section of the HTTP response.

When it comes to serving literal `.gz` files, the actual content of the response is indeed in gzip format, but because there has been no *additional* gzip encoding performed by the HTTP server, setting `Content-Encoding` would be inappropriate. Setting `Content-Encoding: gzip` for `.gz` files would cause browsers to save `.gz` file on disk that has the decompressed data as its contents, rather than being a verbatim copy of the `.gz` file that was requested.

**Comment by mathiasbynens** [comment]
Thanks for clarifying that. My assumptions were overly zealous indeed, as they don’t apply to every file type.

Do you agree it would make sense for e.g. `.css.gz` and `.js.gz` files to get the `Content-Encoding: gzip` header? (Test case: https://mathiasbynens.github.io/gz/)

In the case of the abovementioned project, I wanted/expected `.map.gz` files to be decompressed by the browser. Could `.map.gz` be added to that list, or could we come up with a more generic double extension suffix that signals to GitHub Pages that it should add `Content-Encoding: gzip` (e.g. `.bin.gz`?)

**Comment by mhart** [comment]
I personally wouldn't want that inconsistency – if I wanted ppl on my site to download a large gzipped JS file, then the browser's suddenly going to decompress it and display it inline.

I don't think this should happen transparently – some sort of opt-in would be best. The problem with a generic extension is what `Content-Type` will it be?

**Comment by mathiasbynens** [comment]
> The problem with a generic extension is what `Content-Type` will it be?

A `Content-Type` of `application/octet-stream` would be fine.

For my use case, all that matters is that I can XHR in the decompressed binary data.

**Comment by mhart** [comment]
That's such a specific use case though. It wouldn't work for your JS/CSS case. For example, your server that's serving the `.css.gz` is sending a `Content-Type` of `text/css; charset=utf-8` so that the browser can correctly display it – so that would be incorrect with a `Content-Type` of `application/octet-stream`

**Comment by mathiasbynens** [comment]
@mhart I think we can all agree `.css.gz` and `.js.gz` should get the right `Content-Type`. What’s under discussion now is how to do something similar for generic precompressed binary files. A `Content-Type` of `application/octet-stream` is fine for that use case.

**Comment by mhart** [comment]
@mathiasbynens No, but we disagree even on that :smile: – I *wouldn't* want a `.css.gz` file to have a `Content-Type` of `text/css` by default – for the very reason I outlined above. It's *not* a CSS file.

**Comment by mhart** [comment]
You can explicitly set your server up to serve all `.css.gz` files that way if you want, but that's certainly not standard – and it also means you can't link to a gzipped CSS file if you want ppl to download it that way (eg, "Click here to download my gzipped bootstrap CSS") – the browser will always interpret it as a CSS file, not a gzip file.

**Comment by mathiasbynens** [comment]
@mhart I guess I don’t understand your use case, then. Why would you want people to download a `.css.gz` file and *not* decompress it?

**Comment by mhart** [comment]
Because you don't want the browser to presuppose what ppl want to do with it. It could be huge, for example. It might be a gzip file with a specific compression that you want someone to look at. Again, it's a gzip file. You just happened to have named it in a particular way.

**Comment by mhart** [comment]
The `gzip_static` approach is certainly viable – it allows you to signal some sort of intent – I'm saying it should happen transparently (for *any* extension)

**Comment by TazeTSchnitzel** [comment]
I'd be fine if it adds this header for `.gz` files, but only if that's when you're not requesting the URL of the file itself. If I ask for the `.gz` file, I want that. If I ask for a non-`.gz` file and it doesn't exist, but there is a file with `.gz` appended to its name, then I'd be fine with GitHub serving me that file with this header.

So, if I have a directory with the file `foobar.js.gz`, requesting `/foobar.js.gz` would give me the raw gzipped file with `Content-Type: application/octet-stream`, whereas requesting `/foobar.js` would send me `foobar.js.gz` with the 'Content-Encoding: gzip` and `Content-Type: text/javascript` headers.

**Comment by mhart** [comment]
Yep, I agree 100%.

There's still a tiny amount of room for WTFness (like, I can imagine the odd "why does this particular `.css` file download gzipped and this one doesn't?" question popping up) – but it's a relatively straightforward mechanism.

**Comment by TazeTSchnitzel** [comment]
For the transparent gzip compressed-in-transit case, it may be best to make this only apply if there's only a `.gz` version of the document and not one without that extension as well. Otherwise, you'd be either always ignoring the non-compressed one, or potentially serving a different file to some clients. Have one version of the file, either compressed or decompressed. You can block support for non-gzip-accepting clients or decode for them.

**Comment by mathiasbynens** [comment]
> There's still a tiny amount of room for WTFness (like, I can imagine the odd "why does this particular `.css` file download gzipped and this one doesn't?" question popping up)

How so? GitHub Pages already serves `.css` files with gzip compression enabled.

**Comment by mhart** [comment]
🙄 That makes the entire conversation about `.css.gz` being served as `text/css` even *more* moot then.

Use a different suffix then and you'll see my point.

**Comment by mlknz** [comment]
I have a XHR request with set `Accept-Encoding:gzip` header to `.dds` file and get response with `Content-Type:application/octet-stream` and **without** `Content-Encoding`. `.dds.gz` exists close to it and is not used.

How could I get gzipped file with right response header?

**Comment by Pomax** [comment]
for the purpose of discussion, perhaps replace `css` with `svg`. I have about 2MB in svg images, loaded as images on-page using `<img src="...">` on a gh-pages branch. That's a lot of data, most of which is unnecessary: gzip would compress this optimized, but plain text data to about 500KB instead. As such, having gzip enabled with "the browser gets to decide what to do, based on the content type requested" makes a non-trivial difference.

**Comment by zbeekman** [comment]
Yes, it would be great to be able to serve up gzipped images for faster page loads... not sure if there is a way to put some control in the users hands, or just find a set of sane defaults.

**Comment by Pomax** [comment]
given that jekyll is turned on even for projects that are already their own website content, requiring a `.nojekyll` file to actually stop doing "the default thing", I would certainly advocate simply moving forward with turning on gz for everything, and telling people to use a ".nogzip" if that's not what they want to have happen. 

And of course on the UI side add some options in the repo settings for github pages so that people can set those things without needing files, so a checkbox for "disable jekyll compilation" and "disable hosting with gzip". 

Sane defaults with the option to disable is generally a great way forward.

**Comment by JoshyPHP** [comment]
I'd like to add that ngx_brotli has a similar setting for precompressed brotli content: https://github.com/google/ngx_brotli#brotli_static

It's nice to be able to compress static content at the highest/slowest level offline (perhaps as part of one's deployment procedure) and get excellent compression for free.

**Comment by prideout** [comment]
As a workaround, I've been adding a `.bmp` suffix to any file that I wish to be compressed, even when they are not bitmaps.  In this way I trick nginx into compressing it.

**Comment by MatTheCat** [comment]
Given https://github.community/t5/GitHub-Pages/Support-for-pre-compressed-assets-and-brotli-compression/td-p/22055 this feature should be on the roadmap but I couldn't find anything else.

Does someone know something?

**Comment by vchirikov** [comment]
4 years later...

**Comment by KurtGokhan** [comment]
Any plans to get this implemented?
